golang/pkg/rnr/tasks.go

```diff
@@ -28,10 +33,84 @@ func taskSchedState(pbt *pb.Task) TaskState {
 	return UNKNOWN
 }
 
+// TaskCallback is installed by higher-level construct, such as callback task or nested task.
+type TaskCallback func(context.Context, *Task)
+type StateUpdateCallback func(*pb.Task) *pb.Task
+
 // Task is a generic interface for pollable tasks
-type Task interface {
-	Poll()
-	Proto(updater func(*pb.Task)) *pb.Task
-	SetState(pb.TaskState)
-	GetChild(name string) Task
+type Task struct {
+	cb           TaskCallback
+	pb           *pb.Task
+	children     []*Task
+	has_children bool
+}
+
+func NewTask(name string, children bool, cb TaskCallback) *Task {
+	return &Task{
+		cb: cb,
+		pb: &pb.Task{
+			Name:  name,
+			State: pb.TaskState_PENDING,
+		},
+		children:     []*Task{},
+		has_children: children,
+	}
+}
+
+func (task *Task) Poll() {
+	task.cb(context.TODO(), task)
+}
+
+func (task *Task) Proto(updater StateUpdateCallback) *pb.Task {
+
+	oldState, ok := proto.Clone(task.pb).(*pb.Task)
+	if !ok {
+		log.Fatalf("Failed to clone proto")
```

Comment:
It feels odd that a library calls `log.Fatal`. The best solution would be to return an error here instead, but given the current code, wouldn't be better if it calls `panic` instead of `log.Fatal`? It will produce the same effect of stopping the program but also provide a useful stack trace when something really unexpected happens.

Reply:
I'll be happy with either. The reason for failing hard in a library is that I don't know how the caller should actually handle such an error.

The caller will usually be a Poll() callback of a task or similar that wants to modify a task state. If task state modification fails for any reason, we're basically failing to sync callback state with the proto (i.e. failing to transition from RUNNING to SUCCESS only due to failed proto update).